File: 0004.배열돌리기/solve.1.16926.py
'''
단순 로테이션 방식은 파이썬에서 시간 초과:
C 언어로 구현해서 시간 초과 문제를 해결
파이썬으로는 리스트 슬라이싱 방식을 사용함
'''

def rotate(k, N, M, R, A):
    cycle = []
    # add left line to the bottom
    for i in range(k, N-1-k):
        cycle.append(A[i][k])
    # add bottom line to the right
    for j in range(k, M-1-k):
        cycle.append(A[N-1-k][j])
    # add right line to the top
    for i in range(N-1-k, k, -1):
        cycle.append(A[i][M-1-k])
    # add top line to the left
    for j in range(M-1-k, k, -1):
        cycle.append(A[k][j])
    index = len(cycle) - R % len(cycle)
    rotated = cycle[index:] + cycle[:index]
    t = 0
    # add left line to the bottom
    for i in range(k, N-1-k):
        A[i][k] = rotated[t]
        t += 1
    # add bottom line to the right
    for j in range(k, M-1-k):
        A[N-1-k][j] = rotated[t]
        t += 1
    # add right line to the top
    for i in range(N-1-k, k, -1):
        A[i][M-1-k] = rotated[t]
        t += 1
    # add top line to the left
    for j in range(M-1-k, k, -1):
        A[k][j] = rotated[t]
        t += 1

N, M, R = map(int, input().split())
A = [list(map(int, input().split())) for _ in range(N)]
for k in range(min(N, M) // 2):
    # k is an index from the outer cycle into the inner cycle
    rotate(k, N, M, R, A)
for i in range(len(A)):
    print(" ".join(map(str, A[i])))

File: 0004.배열돌리기/solve.1.16926.c
#include <stdio.h>

#define MAX 300

int A[MAX][MAX];

int min(int a, int b) {
    return (a < b) ? a : b;
}

void rotate(int N, int M) {
    int i, j, k, r, c, n, m, t;
    // rotate from the outer cycle to the inner cycle.
    for (k = 0; k < min(N, M) / 2; k++) {
        r = k; c = k; n = N - 1 - k; m = M - 1 - k;
        t = A[r][c];
        // move top line to the left
        for (j = c; j < m; j++)
            A[r][j] = A[r][j + 1];
        // move right line to the top
        for (i = r; i < n; i++)
            A[i][m] = A[i + 1][m];
        // move bottom line to the right
        for (j = m; j > c; j--)
            A[n][j] = A[n][j - 1];
        // move left line to the bottom
        for (i = n; i > r + 1; i--)
            A[i][c] = A[i - 1][c];
        A[r + 1][c] = t;
    }
}

void solve(int N, int M, int R) {
    int i;
    for (i = 0; i < R; i++)
        rotate(N, M);
}

int main() {
    int N, M, R, i, j;
    scanf("%d %d %d", &N, &M, &R);
    for (i = 0; i < N; i++)
        for (j = 0; j < M; j++)
            scanf("%d", &A[i][j]);
    solve(N, M, R);
    for (i = 0; i < N; i++) {
        for (j = 0; j < M; j++)
            printf("%d ", A[i][j]);
        printf("\n");
    }
}